### ISSUE TITLE
Query.cursor().next(): Maximum call stack size exceeded 

### ISSUE THREAD

**Issue by wellcaffeinated** [Automattic/mongoose#8010]
Bug

I'm getting a strange error when attempting to use a query.cursor() to iterate through documents. I'm not sure what the problem is, because I use the same process to successfully iterate through other models.

I can't really isolate the problem because it works with other models, and even on the current model with a different query... so I'm confused as to what's going on. Any help would be appreciated.

The general idea is that I set up a way to call `cursor.next()` in parallel calls (max 5 concurrent calls) and a small function is run on each document to generate output. **There are no update calls to the database, this is read-only.**

```
/app/node_modules/mongodb/lib/utils.js:132 
       throw err; 
       ^ 
 RangeError: Maximum call stack size exceeded 
     at model.Document.$__init (/app/node_modules/mongoose/lib/document.js:500:8) 
     at model.syncWrapper [as $__init] (/app/node_modules/kareem/index.js:234:23) 
     at model.Document.init (/app/node_modules/mongoose/lib/document.js:466:8) 
     at _create (/app/node_modules/mongoose/lib/cursor/QueryCursor.js:323:12) 
     at /app/node_modules/mongoose/lib/cursor/QueryCursor.js:278:11 
     at result (/app/node_modules/mongodb/lib/utils.js:410:17) 
     at executeCallback (/app/node_modules/mongodb/lib/utils.js:402:9) 
     at handleCallback (/app/node_modules/mongodb/lib/utils.js:128:55) 
     at cursor._next (/app/node_modules/mongodb/lib/operations/cursor_ops.js:195:5) 
     at handleCallback (/app/node_modules/mongodb-core/lib/cursor.js:204:5) 
     at nextFunction (/app/node_modules/mongodb-core/lib/cursor.js:578:5) 
     at Cursor.next (/app/node_modules/mongodb-core/lib/cursor.js:766:3) 
     at Cursor._next (/app/node_modules/mongodb/lib/cursor.js:216:36) 
     at nextObject (/app/node_modules/mongodb/lib/operations/cursor_ops.js:192:10) 
     at next (/app/node_modules/mongodb/lib/operations/cursor_ops.js:171:3) 
     at executeOperation (/app/node_modules/mongodb/lib/utils.js:416:24) 
     at Cursor.next (/app/node_modules/mongodb/lib/cursor.js:258:10) 
     at _next (/app/node_modules/mongoose/lib/cursor/QueryCursor.js:266:23) 
     at utils.promiseOrCallback.cb (/app/node_modules/mongoose/lib/cursor/QueryCursor.js:176:5) 
     at Object.promiseOrCallback (/app/node_modules/mongoose/lib/utils.js:248:12) 
     at QueryCursor.next (/app/node_modules/mongoose/lib/cursor/QueryCursor.js:175:16) 
     at getNext (/app/lib/observable-from-cursor.js:10:27) 
/app/node_modules/mongodb/lib/utils.js:132 
      throw err; 
      ^ 
RangeError: Maximum call stack size exceeded 
    at new ValidatorError (/app/node_modules/mongoose/lib/error/validator.js:23:24) 
    at _init (/app/node_modules/mongoose/lib/document.js:561:37) 
    at init (/app/node_modules/mongoose/lib/document.js:528:5) 
    at model.Document.$__init (/app/node_modules/mongoose/lib/document.js:498:3) 
    at model.syncWrapper [as $__init] (/app/node_modules/kareem/index.js:234:23) 
    at model.Document.init (/app/node_modules/mongoose/lib/document.js:466:8) 
    at _create (/app/node_modules/mongoose/lib/cursor/QueryCursor.js:323:12) 
    at /app/node_modules/mongoose/lib/cursor/QueryCursor.js:278:11 
    at result (/app/node_modules/mongodb/lib/utils.js:410:17) 
    at executeCallback (/app/node_modules/mongodb/lib/utils.js:402:9) 
    at handleCallback (/app/node_modules/mongodb/lib/utils.js:128:55) 
    at cursor._next (/app/node_modules/mongodb/lib/operations/cursor_ops.js:195:5) 
    at handleCallback (/app/node_modules/mongodb-core/lib/cursor.js:204:5) 
    at nextFunction (/app/node_modules/mongodb-core/lib/cursor.js:578:5) 
    at Cursor.next (/app/node_modules/mongodb-core/lib/cursor.js:766:3) 
    at Cursor._next (/app/node_modules/mongodb/lib/cursor.js:216:36) 
    at nextObject (/app/node_modules/mongodb/lib/operations/cursor_ops.js:192:10) 
    at next (/app/node_modules/mongodb/lib/operations/cursor_ops.js:171:3) 
    at executeOperation (/app/node_modules/mongodb/lib/utils.js:416:24) 
    at Cursor.next (/app/node_modules/mongodb/lib/cursor.js:258:10) 
    at _next (/app/node_modules/mongoose/lib/cursor/QueryCursor.js:266:23) 
    at utils.promiseOrCallback.cb (/app/node_modules/mongoose/lib/cursor/QueryCursor.js:176:5) 
```

Mongoose: 5.5.4
Node: 10.15.3
Mongodb: 3.6.11

**Comment by vkarpov15** [comment]
Unfortunately you can't call `next()` on the same cursor multiple times in parallel: https://jira.mongodb.org/browse/NODE-926 .

That being said, this stack trace is very strange, it looks like it is actually getting a document, just Mongoose is crashing when it is trying to instantiate the document. I'll see if I can repro this. Can you show what your schema looks like?

**Comment by wellcaffeinated** [comment]
schema looks like this, but the bug only happened for specific documents... ~800 count query result

```js
new mongoose.Schema( {
        code: {
            type: String
            , validate: [ _.partialRight( _.result, 'length' ), 'Code can not be blank' ]
            , required: true
            , set: function( v ){ return v.toUpperCase(); }
        }
        , rules: {
            type: [ {
                value: Number
                , type: {
                    // percent, flat, fixed
                    type: String
                    , validate: [ function ( val ) {
                        return val === 'percent' ||
                            val === 'flat' ||
                            val === 'fixed';
                    }, 'Coupon type must be one of: "percent", "flat", or "fixed"' ]
                    , required: true
                }
                , ticket: {
                    type: mongoose.Schema.Types.ObjectId
                    , ref: 'Ticket'
                    , validate: [ _.partialRight( _.result, 'length' ), 'Coupon rule must be associated with a ticket' ]
                    , required: true
                }
                , limit: {
                    type: Number
                    , min: 0
                }
                , counts: {
                    used: {
                        type: Number
                        , min: 0
                        , default: 0
                    }
                }
            } ]
            , validate: [ _.partialRight( _.result, 'length' ), 'Coupon rules cannot be empty' ]
            , required: true
            , set: _.toArray /* ensure this is saved as an array */
            , default: []
        }
        , expires: {
            type: Date
            , set: function( val ){
                if (!val) { return; }
                return val;
            }
        }
        , event: {
            type: mongoose.Schema.Types.ObjectId
            , ref: 'Event'
            , validate: [ _.partialRight( _.result, 'length' ), 'Event can not be blank' ]
            , required: true
        }
        , notes: String
        , tags: [{}]
        , created: {
            type: Date
            , default: Date.now
        }
    } );
```

**Comment by wellcaffeinated** [comment]
> Unfortunately you can't call `next()` on the same cursor multiple times in parallel: https://jira.mongodb.org/browse/NODE-926 .

And... that's weird... i've been successfully running queries in parallel, i thought...

**Comment by vkarpov15** [comment]
Can you show me how you're iterating through the cursor? The below script errors out immediately for me:

```javascript
const assert = require('assert');
const mongoose = require('mongoose');

const GITHUB_ISSUE = `gh8010`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;
const { Schema } = mongoose;

run().then(() => console.log('done')).catch(error => console.error(error.stack));

async function run() {
  await mongoose.connect(connectionString);
  await mongoose.connection.dropDatabase();

  const schema = new Schema({ name: String, number: Number });
  const Model = mongoose.model('Test', schema);

  let name = '';
  while (name.length < 10000) {
    name += 'a';
  }

  for (let i = 0; i < 5000; ++i) {
    await Model.create({ name, number: i });
  }

  const cursor = Model.find().cursor();

  while (true) {
    const docs = await Promise.all([
      cursor.next(),
      cursor.next(),
      cursor.next(),
      cursor.next()
    ]);

    if (docs.filter(doc => doc != null).length < docs.length) {
      break;
    }

    for (const doc of docs) { console.log(doc.number); }
  }
}
```

Output

```
$ node gh-8010.js 
(node:30091) DeprecationWarning: current URL string parser is deprecated, and will be removed in a future version. To use the new parser, pass option { useNewUrlParser: true } to MongoClient.connect.
Getting docs
/home/val/Workspace/MongoDB/troubleshoot-mongoose/node_modules/mongodb/lib/utils.js:132
      throw err;
      ^

TypeError: Cannot read property 'high_' of null
    at Long.equals (/home/val/Workspace/MongoDB/troubleshoot-mongoose/node_modules/bson/lib/bson/long.js:236:31)
    at nextFunction (/home/val/Workspace/MongoDB/troubleshoot-mongoose/node_modules/mongodb-core/lib/cursor.js:547:15)
    at Cursor.next (/home/val/Workspace/MongoDB/troubleshoot-mongoose/node_modules/mongodb-core/lib/cursor.js:768:3)
```

That's not a very good error message, but it is indicative of the fact that you shouldn't be calling `next()` multiple times in parallel.

**Comment by vkarpov15** [comment]
I opened up an issue here: https://jira.mongodb.org/browse/NODE-2083

**Comment by vkarpov15** [comment]
Appears fixed as of `mongodb@3.4.1`
